src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+        x = x.transpose(0, 1)
+
+        return x, all_hidden_states, list(all_self_attns)
+
+    def max_positions(self):
+        """Maximum output length supported by the decoder."""
+        return min(self.max_target_positions, self.embed_positions.max_positions)
+
+    def buffered_future_mask(self, tensor):
+        """Upper triangular matrix filled with negative inf for masking."""
+        # TODO(SS): identical to encoder code
+        dim = tensor.size(0)
+        if (
+            not hasattr(self, "_future_mask")
+            or self._future_mask is None
+            or self._future_mask.device != tensor.device
+            or self._future_mask.size(0) < dim
+        ):
+            self._future_mask = torch.triu(fill_with_neg_inf(tensor.new(dim, dim)), 1)
+        return self._future_mask[:dim, :dim]
+
+
+# Helper Modules
+
+
+class BARTClassificationHead(nn.Module):
+    """Head for sentence-level classification tasks."""
+
+    # This can trivially be shared with RobertaClassificationHead
+
+    def __init__(
+        self, input_dim, inner_dim, num_classes, pooler_dropout,
+    ):
+        super().__init__()
+        self.dense = nn.Linear(input_dim, inner_dim)
+        self.dropout = nn.Dropout(p=pooler_dropout)
+        self.out_proj = nn.Linear(inner_dim, num_classes)
+
+    def forward(self, x, **unused):
+        x = self.dropout(x)
+        x = self.dense(x)
+        x = torch.tanh(x)
+        x = self.dropout(x)
+        x = self.out_proj(x)
+        return x
+
+
+class LearnedPositionalEmbedding(nn.Embedding):
+    """
+    This module learns positional embeddings up to a fixed maximum size.
+    Padding ids are ignored by either offsetting based on padding_idx
+    or by setting padding_idx to None and ensuring that the appropriate
+    position ids are passed to the forward function.
+    """
+
+    def __init__(
+        self, num_embeddings: int, embedding_dim: int, padding_idx: int,
+    ):
+        # if padding_idx is specified then offset the embedding ids by
+        # this index and adjust num_embeddings appropriately
+        assert padding_idx is not None
+        num_embeddings += padding_idx + 1  # WHY?
+        super().__init__(num_embeddings, embedding_dim, padding_idx=padding_idx)
+        self.max_positions = num_embeddings - self.padding_idx - 1
+
+    def forward(self, input, incremental_state=None):
+        """Input is expected to be of size [bsz x seqlen]."""
+        if incremental_state is not None:
+            # positions is the same for every token when decoding a single step
+            # Without the int() cast, it doesn't work in some cases when exporting to ONNX
+            positions = input.data.new(1, 1).fill_(int(self.padding_idx + input.size(1)))
+        else:
+            positions = self.make_positions(input, self.padding_idx)
+        return super().forward(positions)
+
+    @staticmethod
+    def make_positions(tensor, padding_idx: int):
```

Comment:
maybe uniformize with the one that we have in RoBERTa (that was half contributed by someone, half rewritten by us)